Sources/MapboxCoreNavigation/RouteController.swift

```diff
@@ -225,11 +225,9 @@ open class RouteController: NSObject {
     }
     
     /// updateRouteLeg is used to notify nav-native of the developer changing the active route-leg.
-    private func updateRouteLeg(to value: Int) {
-        let legIndex = UInt32(value)
-        if navigator.changeRouteLeg(forRoute: 0, leg: legIndex), let status = navigator.getStatus() {
-            updateIndexes(status: status, progress: routeProgress)
-        }
+    private func updateRouteLeg(to legIndex: Int) {
+        precondition(legIndex >= 0, "Leg index can't be negative")
+        navigator.changeRouteLeg(forRoute: 0, leg: UInt32(legIndex))
```

Comment:
Even if the navigator responds to the leg change and updates the navigator status very quickly and synchronously, RouteController responds to the status update asynchronously. This could violate the developer’s expectation, because `updateRouteLeg(to:)` is called from the public `Router.advanceLegIndex()` method, which looks like a synchronous method to the developer. The developer may be following up with operations that depend on the status being updated, for example to update UI to reflect the new leg.

To set the right expectations, I think `Router.advanceLegIndex()` will need to be renamed `Router.advanceLegIndex(completionHandler:)`. Alternatively, we could leave the redundant `getStatus()` in place if MapboxNavigationNative considers it harmless.